source/credentials_provider_sso.c

```diff
@@ -0,0 +1,864 @@
+        aws_byte_buf_append_encoding_uri_param(&wrapped_user_data->path_and_query, &role_name_cursor)) {
+        goto on_error;
+    }
+
+    if (aws_byte_buf_init(&wrapped_user_data->payload, provider->allocator, SSO_RESPONSE_SIZE_INITIAL)) {
+        goto on_error;
+    }
+    return wrapped_user_data;
+
+on_error:
+    s_user_data_destroy(wrapped_user_data);
+
+    return NULL;
+}
+
+/*
+ * No matter the result, this always gets called assuming that user_data is successfully allocated
+ */
+static void s_finalize_get_credentials_query(struct sso_user_data *user_data) {
+    struct aws_credentials *credentials = NULL;
+    if (user_data->status_code == AWS_HTTP_STATUS_CODE_200_OK && user_data->error_code == AWS_OP_SUCCESS) {
+        /* parse credentials */
+        struct aws_parse_credentials_from_json_doc_options parse_options = {
+            .access_key_id_name = "accessKeyId",
+            .secret_access_key_name = "secretAccessKey",
+            .token_name = "sessionToken",
+            .expiration_name = "expiration",
+            .top_level_object_name = "roleCredentials",
+            .token_required = true,
+            .expiration_required = true,
+            .expiration_format = AWS_PCEF_NUMBER_UNIX_EPOCH_MS,
+        };
+
+        credentials = aws_parse_credentials_from_json_document(
+            user_data->allocator, aws_byte_cursor_from_buf(&user_data->payload), &parse_options);
+    }
+
+    if (credentials) {
+        AWS_LOGF_INFO(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p) successfully queried credentials", (void *)user_data->provider);
+    } else {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p) failed to query credentials", (void *)user_data->provider);
+
+        if (user_data->error_code == AWS_ERROR_SUCCESS) {
+            user_data->error_code = AWS_AUTH_CREDENTIALS_PROVIDER_SSO_SOURCE_FAILURE;
+        }
+    }
+
+    /* pass the credentials back */
+    user_data->original_callback(credentials, user_data->error_code, user_data->original_user_data);
+
+    /* clean up */
+    s_user_data_destroy(user_data);
+    aws_credentials_release(credentials);
+}
+static void s_on_retry_ready(struct aws_retry_token *token, int error_code, void *wrapped_user_data);
+
+static void s_on_stream_complete_fn(struct aws_http_stream *stream, int error_code, void *data) {
+    struct sso_user_data *user_data = data;
+
+    /* set error code */
+    user_data->error_code = error_code;
+    if (error_code == AWS_OP_SUCCESS && user_data->status_code != AWS_HTTP_STATUS_CODE_200_OK) {
+        user_data->error_code = AWS_AUTH_CREDENTIALS_PROVIDER_HTTP_STATUS_FAILURE;
+    }
+
+    struct aws_credentials_provider_sso_impl *impl = user_data->provider->impl;
+    impl->function_table->aws_http_stream_release(stream);
+
+    /*
+     * If we can retry the request based on error response, retry it, otherwise, call the finalize function.
+     */
+    if (user_data->error_code) {
```

Comment:
should we also try to retry based on the status code?

```suggestion
    if (user_data->error_code || user_data->status_code!=AWS_HTTP_STATUS_CODE_200_OK) {
```

Reply:
We are already doing that as `user_data->error_code` will be set if http status code was not 200. I have changed it to your suggestion as it is more clear. thanks.

Reply:
Oh, I missed the part you set the error code based on status code. It's fine, it just couple line above, we can keep what you did before.